examples/list2.c

```diff
@@ -0,0 +1,256 @@
+  match_Cons_0:;
+    ssm_value_t __i = ssm_adt_field(cont->l, 0);
+    ssm_value_t __l = ssm_adt_field(cont->l, 1);
+    ssm_dup(__l);
+
+    ssm_drop(cont->l);
+
+    ssm_dup(__i);
+    ssm_activate(ssm_closure_reduce(cont->inc, __i, act, act->priority, act->depth, &cont->__tmp0));
+    cont->__tmp2 = __l;
+    act->pc = 1;
+    return;
+  case 1:;
+    ssm_activate(
+        enter_map_inc(act, act->priority, act->depth, cont->__tmp2, cont->inc, &cont->__tmp1));
+    act->pc = 2;
+    return;
+  case 2:;
+    *cont->__ret = ssm_new_adt(List_size[Cons], Cons);
+    ssm_adt_field(*cont->__ret, 0) = cont->__tmp0;
+    ssm_adt_field(*cont->__ret, 1) = cont->__tmp1;
+    break;
+  }
+  ssm_leave(&cont->act, sizeof(act_map_inc_t));
+}
+
+typedef struct {
+  ssm_act_t act;
+  ssm_value_t l;
+  ssm_value_t __tmp0;
+  ssm_value_t __tmp1;
+} act_print_list_t;
+
+ssm_stepf_t step_print_list;
+
+ssm_act_t *enter_print_list(ssm_act_t *parent, ssm_priority_t priority,
+                            ssm_depth_t depth, ssm_value_t l) {
+  act_print_list_t *cont =
+      container_of(ssm_enter(sizeof(act_print_list_t), step_print_list, parent,
+                             priority, depth),
+                   act_print_list_t, act);
+  cont->l = l;
+  return &cont->act;
+}
+
+void step_print_list(ssm_act_t *act) {
+  act_print_list_t *cont = container_of(act, act_print_list_t, act);
+  switch (act->pc) {
+  case 0:
+    switch (ssm_tag(cont->l)) {
+    case Nil:
+      goto match_Nil_0;
+    case Cons:
+      goto match_Cons_0;
+    }
+    SSM_ASSERT(0);
+  match_Nil_0:
+    printf("[]\r\n");
+    break;
+  match_Cons_0:;
+    ssm_value_t __i = ssm_adt_field(cont->l, 0);
+    ssm_value_t __l = ssm_adt_field(cont->l, 1);
+    ssm_dup(__l);
+
+    ssm_drop(cont->l);
+
+    printf("%d::", ssm_unmarshal(__i));
+
+    ssm_activate(enter_print_list(act, act->priority, act->depth, __l));
+    act->pc = 1;
+    return;
+  case 1:
+    break;
+  }
+  ssm_leave(&cont->act, sizeof(act_print_list_t));
+}
+
+typedef struct {
+  ssm_act_t act;
+  ssm_value_t list;
+  ssm_value_t list_;
+  ssm_value_t inc_func;
+  ssm_value_t init;
+  ssm_value_t offset;
+} act_main_t;
+
+ssm_stepf_t step_main;
+
+// Create a new activation record for main
+ssm_act_t *ssm_enter_main(struct ssm_act *parent, ssm_priority_t priority,
+                          ssm_depth_t depth) {
+  act_main_t *cont = container_of(
+      ssm_enter(sizeof(act_main_t), step_main, parent, priority, depth),
+      act_main_t, act);
+
+  cont->list = list;
+  return &cont->act;
+}
+
+void step_main(struct ssm_act *act) {
+  act_main_t *cont = container_of(act, act_main_t, act);
+
+  switch (act->pc) {
+  case 0:
+    ssm_dup(cont->list);
+    ssm_activate(enter_print_list(act, act->priority, act->depth, cont->list));
+    act->pc = 1;
+    return;
+  case 1:;
+    cont->init= ssm_new_closure(&enter_inc_offset);
+    cont->offset = ssm_marshal(69);
+    cont->inc_func = ssm_closure_apply(cont->init, cont->offset);
```

Comment:
Do all of these have to be stored in the activation record? @j-hui

Reply:
They don't have to, since they're intermediate values that don't live beyond the lifetime of a single instant. You can just declare local variables for them. Make sure to drop the new closure too.

Reply:
Please see 45f92d